clash-lib/src/Clash/Netlist/Expr.hs

```diff
@@ -0,0 +1,166 @@
+{-|
+  Copyright  :  (C) 2012-2016, University of Twente,
+                    2017     , Myrtle Software Ltd,
+                    2017-2018, Google Inc.
+                    2020-2022, QBayLogic B.V.
+                    2022     , Google Inc.
+  License    :  BSD2 (see the file LICENSE)
+  Maintainer :  QBayLogic B.V. <devops@qbaylogic.com>
+
+  Functions for expression manipulation
+-}
+
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module Clash.Netlist.Expr where
+
+import Control.Monad (zipWithM)
+import Control.Exception (assert)
+import Data.Bits (Bits, testBit, setBit, zeroBits)
+import Data.Foldable (fold)
+import Data.Tree (Tree(..))
+import GHC.Stack (HasCallStack)
+import Clash.Primitives.DSL (tySize)
+import Clash.Netlist.Types
+  ( Size, Bit(..), Expr(..), HWType(..), Literal(..), Modifier(..)
+  , BlackBoxContext(..)
+  )
+
+-- | Turns a constant expression of known bitsize into their
+-- corresponding bitstream representation, arranged as a tree
+-- that corresponds to the structure of the expression.
+--
+-- NOTE: This conversion serves as a best effort approach and can be
+-- considered a hack. Fully featured constant expression evaluation is
+-- not available in clash yet and will replace this implementation
+-- once it is officially supported.
+bits :: HasCallStack => Size -> Expr -> Either Expr (Tree [Bool])
+bits size expr = case expr of
+  Literal _ lit -> case lit of
+    BitLit bLit   -> case bLit of
+      H -> leaf [True]
+      L -> leaf [False]
+      _ -> Left expr
+    BoolLit bLit      -> leaf [bLit]
+    NumLit nLit       -> leaf $ toBits size nLit
+    BitVecLit _ bvLit -> leaf $ toBits size bvLit
+    VecLit lits       ->
+      mapM (bits (size `div` length lits) . lit2Expr) lits >>= inner
+    StringLit{}       -> Left expr
+  DataCon ty m subs -> assert (tySize ty == size) $ case ty of
+    Vector s t      -> vecBits (tySize t) s subs
+    Product _ _ tys -> zipWithM bits (map tySize tys) subs >>= inner
+    Sum _ cs        -> spBits expr size m subs $ map (const []) cs
+    SP _ xs         -> spBits expr size m subs $ map (map tySize . snd) xs
+    _               -> case subs of
+      [e] -> bits size e
+      []  -> leaf []
+      _   -> Left expr
+  -- appears in case of complex transformations, e.g.,
+  -- >>> (bv2v 0b010) :: Vec 3 Bit
+  -- >>> (iterate (SNat @3) not True) :: Vec 3 Bool
+  -- >>> (complement <$> (True :> False :> Nil)) :: Vec 2 Bool
+  Identifier{} -> Left expr
+  DataTag{} -> Left expr
+  BlackBoxE bbName _ _ _ _ bbCtx _ -> case bbName of
+    "Clash.Sized.Internal.BitVector.low" -> leaf [False]
+    "Clash.Sized.Internal.BitVector.high" -> leaf [True]
+    "Clash.Sized.Vector.replicate" -> case bbInputs bbCtx of
```

Comment:
If we encounter this we should instruct the user to use `$(lift ...)` for the constants. There's a bunch of these blackboxes, so no need to single out `replicate`.

Reply:
If there's some other reason this is here that I'm missing, please add a note :).

Reply:
I tested all intuitive ways of constructing vectors with methods from `Clash.Sized.Vector` and `replicate` was the only one causing a problem here. It's also on of the methods, that will be mostly used when creating many probes using a vector. So, in my opinion, we should keep it to be more user friendly. The problem with `$(lift ...)` is that most of the user's won't understand why they need to add it and how it solves the problem. Therefore, I perfer not suggesting it as a solution, unless it's really necessary.